### DISCUSSION TITLE
[RFC] OpenAPI aka swagger support

### DISCUSSION THREAD

**Discussion by KATT** [trpc/trpc#271] [Ideas]
Would be possible for each query/mut to have a `schema:`-prop next to the resolver where one could define an api schema of which the structure could be statically enforced as well.

Maybe the input arg of the schema could also be used as an actual input validator?

---

Another way could possibly be to statically infer the swagger schema with codegen, but then one couldn't actually _document_ properties.

**Comment by transitive-bullshit** [comment]
Love this project so much 💪

Just wanted to link to a few related projects which you might find helpful for adding OpenAPI support:
- https://github.com/tiangolo/fastapi
- https://github.com/lukeautry/tsoa
- https://github.com/transitive-bullshit/functional-typescript

**Reply by KATT** [reply]
https://github.com/trpc/trpc/issues/1724

**Comment by mmkal** [comment]
@KATT one thought - could this rely on [zod-to-json-schema](https://www.npmjs.com/package/zod-to-json-schema) as discussed in #714?

If going with this, it might be nice to add an (optional) `output` property on each procedure, that works roughly the same as `input`, but validates that server responses conform to the contract rather than input requests. If using zod, that would be enough to convert to json-schema using zod-to-json-schema and inspect `appRouter._def` to publish a pretty complete OpenAPI definition document as json - no codegen needed (my team are working on doing the input side of this to integrate our server with retool).

An aside: I've mulled requesting an `output` field _anyway_, since using type inference alone is amazing for getting something written up fast, but once procedures stabilise, it'd be nice to have a clear contract that the server is adhering to in the traditional sense. Another benefit is you could have self-documenting refinements that aren't really possible at compile time like `output: z.object({ count: z.number().int().positive() })`

**Comment by KATT** [comment]
Locking this discussion to keep it within https://github.com/trpc/trpc/issues/755 - @transitive-bullshit, feel free to jump into the chat over there!

**Comment by jlalmes** [comment]
For tRPC to OpenAPI compliant endpoints & Swagger schema generation, please use `trpc-openapi`: https://github.com/jlalmes/trpc-openapi
